### ISSUE TITLE
Fix, then add CI for, test suite at higher processor counts

### ISSUE THREAD

**Issue by roystgnr** [idaholab/moose#9430]
### Description of the enhancement or error report

While investigating the remaining #1500 DistributedMesh failures, I discovered that a few tests in the MOOSE suite fail for me at high processor counts despite using ReplicatedMesh.  We should get these fixed, and then bump up the processor count of our Civet tests to avoid future failures.

### Rationale for the enhancement or information for reproducing the error

```METHOD=dbg ./run_tests -p 12 --re 'transfers/multiapp_variable_value_sample_transfer.monomial_to_sub_pp$|outputs/csv.no_time$|outputs/csv.transient$'```
shows EXODIFF or CSVDIFF failures on all three tests for me.  @permcody managed to trigger failures as well, but at different processor counts, so there might be some system-specific undefined behavior or roundoff error here.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

I don't yet understand the failures in the csv tests, but my investigations so far suggest that there might be a subtle underlying bug that might have a wider impact and/or might be hard to fix.

The multiapp failure looks like it might just be a false positive from a poor exodiff tolerance as in #9414 - I'm testing that theory now and if it's correct the fix will be quick.

**Comment by roystgnr** [comment]
This should be ready to go, but I'm not sure how to make a PR to accomplish it.  Are the MOOSE Civet recipes in a private repo?

**Comment by brianmoose** [comment]
Civet recipes are in a private repo, idaholab/civet_recipes.
Usually people just tell me what they want and I add it.

**Comment by roystgnr** [comment]
On the one hand, we'd like to start using as many CPUs as possible in testing, because the more we slice and dice problems the more chance we have of finding bugs in parallel execution.  On the other hand, we're looking at a thousand tests, none of which is on a big enough mesh to scale well at all, so the more CPUs we throw at them the more CPU power we're wasting.

What's the state of the INL build farm?  If you had unlimited CPU to play with then I'd love to get every test run with -np 32 instead of just -np 2.  If CPU time is a factor then we'd need to figure out where the knee of the cost/benefit curve is.  Probably somewhere between -np 3 (which is already an improvement; there have been bugs in the past that triggered on 3 processors but not on 2) and -np 12 (which IIRC was how many it took to trigger the most-difficult-to-trigger parallel-only libMesh bug, many years ago).

**Comment by brianmoose** [comment]
It depends on when you want the testing to happen. If it is OK to do these sort of tests at night when the build machines aren't busy, then we can go wild and do each 2,..,32.
If you want this tested on each PR then, as you say, we need to be a bit more concerned about time. Are you suggesting that we would have multiple tests, -p 2, -p 3, ...? Or just one with a >2 CPU count?

If you want a nightly test(s), I could add that right now as that shouldn't really affect anybody.
If you want to change the PR testing then I would wait to see what @permcody has to say about it.

**Comment by roystgnr** [comment]
We definitely don't want to bother with multiple tests on every PR.  We currently have a test run with -p 2; I would just switch that to -p 8 (or maybe -p 4 if load is a big issue or -p 16 if it isn't).

I really like the idea of doing a nightly sweep through processor counts in addition, though!  Code which works on 8 processors but fails on 32 is extremely rare, so we probably shouldn't bother testing it on every PR, but such code does exist, and testing for it after the fact sounds like a great compromise.  I'd want to be alerted immediately when such tests regress.  Automatic github issue with me tagged?  Email?  Smoke signal?

**Comment by brianmoose** [comment]
We currently don't have anything setup for emailing, etc when something breaks on master. Usually I or somebody else who looks at the CIVET page too much notices the problem.

As far as PR testing goes, there is only a single CIVET client running on each machine and it can use all the CPUs. However, running with higher -p <count> will increase the overall test time since the MOOSE test harness will schedule fewer tests at the same time. On some of the slower build machines, running with -p 8 would essentially resort to running each test one at a time since there we use "-j 12 -l 12".

**Comment by friedmud** [comment]
I could see bumping the default for PRs to `-p 3` or maybe `-p 4` depending on how much that slows things down.  I wouldn't really want it to go higher than that because round-off issues and issues with pre-conditioner parallelization could start creating diffs.  Could cause quite a bit of trouble for someone putting in a new PR.

@roystgnr are you saying that you can currently run the whole test suite with `-p 32` and have everything pass?

**Comment by roystgnr** [comment]
@brianmoose - in that case, would -p 12 be the optimal thing to do for the nightlies?  I believe that would catch literally 98% of the parallel bugs I've ever seen, whereas -p 2 catches more like 50%.

@friedmud - you're right that round-off and preconditioning can create some annoying diffs!  That was part of the problems fixed in #9416, and #9434, and #9489, and #9496, and #9523, and #9541... so at this point I'd say any new exodiff PR has about a 1% chance of failing when re-run on a radically different processor count.  If we habitually turn on use_old_floor we could probably reduce that to about 0.2%.  I don't think we could get any better than that without either adding new options to exodiff or writing our own diff tool.  (Something like https://grinsfem.github.io/grins/html/generic__solution__regression_8C_source.html but using relative rather than absolute tolerances)

I can currently run (test/; I haven't yet tried modules/) with -p 12 and have everything pass.  I'm doing my testing on a 24 core machine shared with two others, so I've been trying to keep some cores free, but I'll try -p 24 again next.  I have run -p 24 within the last couple months, and I've since fixed everything that broke that time (including #9436, which only triggers on 19 or more processors), but since I'm not using -p 24 frequently I wouldn't swear there haven't been any regressions elsewhere.

**Comment by brianmoose** [comment]
@roystgnr Yes we could do -p 12 . That is the lowest number of CPUS we have on the build boxes. Some have more. 
I created a recipe that does 2-12 to get a feel for the times involved (it got picked up by a faster machine that uses -j 16) https://www.moosebuild.org/job/103285/
You can also manually add the recipe to PRs. 
Looks like there are failures for >2

**Comment by roystgnr** [comment]
Nice, thanks!

The ```utils/spline_interpolation.bicubic_spline```, ```executioners/eigen_executioners.test_inverse_power_method```, and ```auxkernels/grad_component.test``` failures look like obvious cases of #9414, but I can't replicate them so I'll have to fix them blindly and hope they'll work on the next sweep.  The ```materials/material.throw_material```, ```time_integrators/implicit-euler.test```, ```auxkernels/(constant|function)_scalar_aux.test```, ```misc/exception.parallel_exception_(jacobian|residual)_transient```, and ```multiapps/multilevel.time_dt_from_master``` exodiffs all show up for me on 24 processors, so I'll fix them there and hopefully that'll be robust enough to fix it on your systems too.  Only the first of those isn't obviously a #9414 case; the rest ought to be quick fixes.

It looks like "-p 8" would be the sweet spot on that 16 core machine, but that would leave 4 cores idling on your 12 core machines, so switching to "-p 4" for PRs (once I fix those exodiffs) is the way to go.  We ought to just do a sweep for nightly builds (weekly builds? this doesn't need to be hit very often, just more often than "whenever Roy thinks to do it manually"), either up to 12 if you can't guarantee a 16 core machine or up to 16 if you can.

**Comment by brianmoose** [comment]
FYI, I updated the recipe to do 2-16 and only run on machines that have 16 cores.
https://www.moosebuild.org/job/103471/
Looks like there was some new failures and a few of the previous tests that used to exodiff now crash (due to thrown exception).

**Comment by roystgnr** [comment]
Yeah, I can replicate a few crashes at 16 processors of code that passed (after the changes in #9569) at 15.  It sounds like the bigger a sweep we can do periodically, the better...

**Comment by roystgnr** [comment]
Okay - as of 2 minutes ago, moose/devel ought to be able to pass test/ sweeps from 1 to 16 processors, with any combination of --distributed-mesh and --recover involved.

So, can we get something to scream bloody murder if that ever reverts, and then close this issue?  My inclination would be:

1) Nightly sweeps from 1 to 16 with all four combinations of the above options, in devel mode, with *some* automatic notification (Github issue auto-re-opened, email to me, whatever) if there are any failures.
2) Change the current per-PR --distributed-mesh test from -p 2 to -p 4; that's probably the best compromise we can get between speed and test coverage.

**Comment by brianmoose** [comment]
We are already doing the various sweeps on the nightly testing. There isn't any automatic notification but we usually know when it breaks.

I will update the recipes to use --distributed-mesh -p 4

**Comment by roystgnr** [comment]
Thanks!  Is there a good URL to use to look at the nightly results?  I presume they're on the https://www.moosebuild.org/branch/1/ list somewhere but if so they're buried by PR tests...

**Comment by brianmoose** [comment]
There is always https://www.moosebuild.org/scheduled/
which is under "Links/Scheduled Events" in the upper right on CIVET.

**Comment by roystgnr** [comment]
Thanks!  I think we're all set, then!
